### PR TITLE
MSC3383: Include destination in X-Matrix Auth Header

### PR DESCRIPTION
**Pull request by jcgruenhage**
[Rendered](https://github.com/Famedly/matrix-doc/blob/jcgruenhage/fed-auth-destination/proposals/3383-fed-auth-destination.md)
[FCP Proposal](https://github.com/matrix-org/matrix-doc/pull/3383#issuecomment-1026774033)

Signed-off-by: Jan Christian Grünhage <jcgruenhage@famedly.com>

Implementation: https://github.com/matrix-org/synapse/pull/11398

Preview: https://pr3383--matrix-org-previews.netlify.app

### CODE REVIEW

**Review by kevincox — APPROVED** [review]
Making vhosting easier sounds great for me.

**Comment by jcgruenhage** [comment]
While working on a synapse patch, I noticed #2195, which might cause some trouble here. I'd suggest that `destination_is` is dropped in favor of using `destination` for ISs as well.

**Comment by daenney** [comment]
Conceptually the purpose of this sounds rather similar to what TLS SNI does, but for presenting the right certificate. Since the SNI  value would have to be the `server_name` isn't that enough to solve this use case?

File: proposals/3383-fed-auth-destination.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# MSC3383: Include destination in X-Matrix Auth Header
+
+Currently, a federation request can't be validated mid-flight without talking
+to both the originating server and the destination server. The originating
+server needs to be asked for it's key, and the destination server for it's
```
Comment (DMRobertson) [inline proposals/3383-fed-auth-destination.md:5]: nit: "it's" stands for either "it is" or "it has". The possessive form is always "its" with no apostrophe.

```suggestion
server needs to be asked for its key, and the destination server for its
```

Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md]: Thanks! Resolved and pushed

**Review by ShadowJonathan — APPROVED** [review]
vhosting sounds amazing tbh

**Comment by jcgruenhage** [comment]
Implementation is done, as far as I'm concerned. What's still needed to move this forward?

**Comment by uhoreg** [comment]
> Implementation is done, as far as I'm concerned.

Is https://github.com/matrix-org/synapse/pull/11398 the implementation?

**Comment by jcgruenhage** [comment]
> > Implementation is done, as far as I'm concerned.
> 
> Is [matrix-org/synapse#11398](https://github.com/matrix-org/synapse/pull/11398) the implementation?

Yep, it is

**Comment by turt2live** [comment]
This is suggested by the author to be ready for FCP.

**Comment by jcgruenhage** [comment]
> This is suggested by the author to be ready for FCP.

So, can we get an FCP then? Or does #2195 need to be addressed first?

**Comment by richvdh** [comment]
!mscbot fcp merge

**Comment by richvdh** [comment]
@mscbot fcp merge

File: proposals/3383-fed-auth-destination.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# MSC3383: Include destination in X-Matrix Auth Header
```
Comment (erikjohnston) [inline proposals/3383-fed-auth-destination.md:1]: How do we deal with backwards compatibility? Even if we mandate this for newer spec versions existing software is still going to have to handle `Authorization` headers that don't have it, is that a problem?

**Review by erikjohnston — COMMENTED** [review]
I think this makes sense

File: proposals/3383-fed-auth-destination.md (hunk 2 above)
Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:1]: The simple part: As for the case where the field isn't needed, it'll just work like it works now. Conduit, Dendrite and Synapse all support receiving headers with this included already. They'll simply ignore it during the parsing.

The tougher part is deciding when a server can start relying on the information in this header, because if the server requires that this field is included in the request, older deployments won't be able to federate to these specific servers anymore.

I think this is not a big problem though, considering that this MSC doesn't suggest enforcing that only auth headers including the destination are considered valid. Adding this now paves the way for doing that at some point in the future though.

Comment (erikjohnston) [reply proposals/3383-fed-auth-destination.md:1]: Yup, though I'm a bit wary of having such "optional" features. If you can't rely on having them are they of any use? It'd be nice if we had *some* plan of enforcing this, but I'm not sure how that would work TBH.

Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:1]: I have updated the wording of the MSC to explicitly mention newer implementations receiving requests from older implementations which aren't sending the `destination` field.

As for whether they are of any use: Definitely. In the case of a "SHOULD" requirement for gracefully handling the absence of the destination field, a user can weigh that against their own priorities, and if they do need that proxy in between they can decide against that requirement, while understanding the implication that this means federation with older deployments which haven't been updated will break.

File: proposals/3383-fed-auth-destination.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+`server_name`, because the `Host` header does not necessarily contain the
+`server_name` of the destination server, if delegation via `.well-known` is
+being used.
+
+This hasn't been a problem so far, as the `server_name` of the destination is
+usually known when validating the `Authorization` header, it's the
+`server_name` of the matrix server that's doing the validation. But there's two
+scenarios where this might not be the case: Forward proxies (that act as an API
+gateway for enforcing additional rules), or Matrix Homeservers implementing
+vhosting and have multiple `server_name`s pointing to the same `Host`.
```
Comment (turt2live) [inline proposals/3383-fed-auth-destination.md:15]: the problem statement isn't super clear to me: why isn't the `Host` header possible, specifically? Hosting providers, media repos, custom key servers, etc all currently operate just fine without the added destination flag, which includes vhosting and `.well-known` delegation.

**Comment by turt2live** [comment]
@mscbot concern Unclear on the problem being solved

File: proposals/3383-fed-auth-destination.md (hunk 3 above)
Comment (ShadowJonathan) [reply proposals/3383-fed-auth-destination.md:15]: Host headers also have effects on reverse proxies, and do not specifically tell which server to address, if multiple `.well-known`s from different domains point at this one "landing" domain, the server must either have to "map" a wildcard subdomain to normal domains, or have to guess.

Thus, differentiating on the `Host` header is impractical and hack-y. It *is* possible to work with `Host`, but it error-prone, most likely requires wildcard certificates (for the reverse proxy) for it to work, requires custom software to map the right domain prefix to the right "destination", and requires the domains from which `.well-known`s were set up to have those right domain prefixes attached.

This is in contrast with this solution, where the `Host` is untouched and unaltered, allowing for a single-host certificate to do the job, while the entire flow of discovering and finding the right federation endpoint is defined within the spec, which allows for implementations to tune to one approach.

Comparatively, vHosting and other solutions become easier to set up and manage, and allows little room for error or undefined/unspecced behaviour for this application.

(PS: I'm just going about this from my own impression of the MSC, the author may have different opinions or arguments; mine is effectively "it makes a ton of things easier without resorting to hacks")

Comment (turt2live) [reply proposals/3383-fed-auth-destination.md:15]: It's *more* of a hack to have the proxy parse and understand *some* parts of a header though? Arguably it's also a hack that multiple servers are delegating to a single entry point.

Comment (ShadowJonathan) [reply proposals/3383-fed-auth-destination.md:15]: The reverse proxy is not meant to understand the matrix-specific layer here, only forward the domain to a specific HTTP server (we already "abstract" the server domain from the federation endpoint; they're not necessarily the same endpoint as backed by DNS).

vHosting in general could be considered a hack, yes, but i think that it's not necessarily fair, given the internet conglomerates around central points for multiplexed services (think AWS-hosted services, the whole deal with IP in the first place, and then why reverse proxies exist to de-mux incoming traffic to a single IP), because of the scalability and deduplication it provides.

Matrix vHosting, at sufficient scale and with specialized software, may actually be more optimized than single servers hosting an entire stack by themselves. 

(Though that aspect could create an argument to *not* propagate this idea; to not let matrix hosting companies capture so much of the matrix ecosystem. But that'd call into question the already-existing hosting companies, and is in my opinion only looking to one side of the coin;)

<details>
<summary>*(unrelated argumentation about matrix hosting)*</summary>

Hosting companies (and cloud companies, to a degree) amass monopoly by having lock-in, either soft lock-in (everyone's doing it, come join!), or hard-lock in (AWS' dozens of proprietary services, with proprietary APIs).

To a degree, hosting companies can hold your data hostage, so you'd *have* to use them, but if matrix would allow for a degree of freedom between hosting services (and your own hardware), it'd allow for a rich and healthy model and dynamic of people taking their servers where they want, and never being trapped or "locked" into one specific configuration.

P2P matrix can cover a large area of this, and in the past i've tried to define such a data format you can "takeout" and "submit" into these providers, potentially.

vHosting, in such an environment, wouldn't then really be a scourge or a point of centralization, but rather a cost-saving measure of which people can make their own trade-offs, and can change their mind at any time.

</details>

In short, I think leaving the `Host` header to its own layer where it is (the https "layer"), and adding extra data on top to interpret the "matrix" layer, seems like a proper separation of concerns, and in line with how multiplexing has been done on the web.

Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:15]: > why isn't the Host header possible, specifically?

Because it contains the resolved server name, which is not the one that is used in signature sent in this header. This mapping is not required to be unique by the spec so far, multiple servers resolving to a single host that is not using a single, non-wildcard certificate, should be possible.

> Hosting providers, media repos, custom key servers, etc all currently operate just fine without the added destination flag, which includes vhosting and .well-known delegation.

As I've stated in the alternatives section of the MSC [here](https://github.com/Famedly/matrix-doc/blob/jcgruenhage/fed-auth-destination/proposals/3383-fed-auth-destination.md#alternatives), I'm very aware that this is possible to work around in other ways. I just think that the workarounds are somewhat cumbersome and requiring any data not included in the request for validating it's integrity is a bit weird.

For the usecase that motivated this: A proxy in between that does policy validation needs to ask the destination server taken from the Host header where it can find the server_name from `/_matrix/key/v2/server`. This could be cached, which means it's not too bad, but there is no reason to require this addition step.

If you have any other suggestions for how this could be improved, feel free to share.

Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:15]: > It's _more_ of a hack to have the proxy parse and understand _some_ parts of a header though? Arguably it's also a hack that multiple servers are delegating to a single entry point.

Ftr, this is example is not a regular http proxy for routing/load balancing. This is specifically for a gateway style proxy that enforces policies on the requests passing through.

Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:15]: @turt2live Are your concerns cleared up by this?

**Review by anoadragon453 — APPROVED** [review]
Overall this looks fine to me. I'd just like a bit more explicit clarification for implementation behaviour on the verifying side.

File: proposals/3383-fed-auth-destination.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# MSC3383: Include destination in X-Matrix Auth Header
+
+Currently, a federation request can't be validated mid-flight without some
```
Comment (anoadragon453) [inline proposals/3383-fed-auth-destination.md:3]:
```suggestion
Currently, a federation request can't be [validated](https://spec.matrix.org/v1.2/server-server-api/#request-authentication) mid-flight without some
```

just to make it obvious which bit of the spec contains the problem area.

**Review by turt2live — APPROVED** [review]
thanks for clarifying the MSC! I don't have super strong opinions on this, but it does indeed seem like a good thing to have for certain server architectures.

File: proposals/3383-fed-auth-destination.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+## Proposal
+
+In addition to the currently present `origin`, `key` and `sig` fields, the
+`Authorization` header of the scheme `X-Matrix` used in the Matrix S2S API MUST
+also include a `destination` field, which contains the `server_name` of the
+Matrix Homeserver that the request is being sent to.
+
+A Matrix Homeserver receiving a request over the S2S API SHOULD gracefully
+handle requests that do not include the `destination` field in the
+`Authorization` header for backwards compatibility.
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:48]: Not sure I follow, can you give a bit more details?

[the remark this answers is not in the document]
Comment (erikjohnston) [reply proposals/3383-fed-auth-destination.md:50]: The only new case here is if `destination` doesn't match the homeserver's `server_name` (if they do match then the existing check will fail and the spec says to `401` the request). 

It's fairly clear to me that a homeserver should `401` a request if the `destination` field doesn't match the server's `server_name`? Obviously that check won't happen on existing servers that don't understand `destination`, but I don't think that is a problem.

[the remark this answers is not in the document]
Comment (jcgruenhage) [reply proposals/3383-fed-auth-destination.md:50]: https://github.com/matrix-org/matrix-spec-proposals/commit/58da396b47da7ca7ea18ad958dec8e43e2016371 <- Does this look good?

[the remark this answers is not in the document]
Comment (anoadragon453) [reply proposals/3383-fed-auth-destination.md:50]: That looks good to me! I don't think this needs to be a MUST in this instance.

[the remark this answers is not in the document]
Comment (erikjohnston) [reply proposals/3383-fed-auth-destination.md:50]: Thanks!

**Comment by uhoreg** [comment]
Spec PR: https://github.com/matrix-org/matrix-spec/pull/1067

**Comment by turt2live** [comment]
Merged 🎉
